File: public/content/scripts/EvalScenarioParser.java
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;


public class EvalScenarioParser {

    static class EvaluationScenario {
        public List<Integer> Agents;
        public List<Object> Anchors;

        @JsonProperty("True Positions of Agents")
        public Map<String, Position> truePositionsOfAgents;

        @JsonProperty("True Positions of Anchors")
        public Map<String, Position> truePositionsOfAnchors;

        public List<Iteration> Iterations;

        public static class Position {
            public double x;
            public double y;
            public double z;
        }

        public static class Iteration {
            public int ID;

            @JsonProperty("Initial positions of Agents")
            public Map<String, Position> initialPositionsOfAgents;

            @JsonProperty("Ranging samples between Agents")
            public Map<String, Map<String, Object>> rangingSamplesBetweenAgents;

            @JsonProperty("Ranging samples to Anchors")
            public Map<String, Map<String, Object>> rangingSamplesToAnchors;
        }

        // A method to get an Iteration by ID
        public Iteration getIteration(int iteration_id){
            for (Iteration iteration: Iterations){
                if (iteration.ID == iteration_id){
                    return iteration;
                }
            }
            return null; // Return null if the iteration with the given ID is not found
        }
    }

    // Path to the JSON file containing the evaluation scenario data.
    static String dbFilePath = "/Users/geogouz/Desktop/eval scenarios/Local/TWR-CLOUD Bern Eval Scenarios/6905.json";

    /**
     * Parses the evaluation scenario JSON file and outputs the node configuration details.
     * <p>
     * This method reads a JSON file containing information about nodes (agents and anchors) in an evaluation scenario.
     * It parses the file to extract and print:
     * <ul>
     *     <li>Lists of agents and anchors in the scenario</li>
     *     <li>True positions of agents and anchors</li>
     *     <li>Initial positions and ranging samples for each iteration in the scenario</li>
     * </ul>
     * </p>
     * <p>
     * Each parsed entry is displayed in the console, with structured data printed for each entity and iteration.
     * This allows for easy verification of node setup details, including positions and sample data.
     * </p>
     *
     * @throws Exception if there is an error during file reading or parsing.
     */
    static void getNodeSetupFromJsonDB() {
        try {
            System.out.println("Parsing Node JSON DB");

            // Read the entire content of the file into a String
            String jsonString = Files.readString(Path.of(dbFilePath));

            ObjectMapper mapper = new ObjectMapper();
            EvaluationScenario evaluationScenarioDatabase = mapper.readValue(jsonString, EvaluationScenario.class);

            System.out.println("\nNodes in scenario:");
            System.out.println(evaluationScenarioDatabase.Agents);
            System.out.println(evaluationScenarioDatabase.Anchors);

            System.out.println("\nTrue Positions of Agents");
            for (Map.Entry<String, EvaluationScenario.Position> entry :
                    evaluationScenarioDatabase.truePositionsOfAgents.entrySet()) {
                String nodeId = entry.getKey();
                EvaluationScenario.Position nodePos = entry.getValue();
                System.out.println(
                        "Agent: " + nodeId + " x: " + nodePos.x + ", y: " + nodePos.y + ", z: " + nodePos.z
                );
            }

            System.out.println("\nTrue Positions of Anchors");
            for (Map.Entry<String, EvaluationScenario.Position> entry :
                    evaluationScenarioDatabase.truePositionsOfAnchors.entrySet()) {
                String nodeId = entry.getKey();
                EvaluationScenario.Position nodePos = entry.getValue();
                System.out.println(
                        "Anchor: " + nodeId + " x: " + nodePos.x + ", y: " + nodePos.y + ", z: " + nodePos.z
                );
            }

            // Get the iteration data, holding all measurements and initializations.
            for (int iteration = 1; iteration <= 2; iteration++) {
                EvaluationScenario.Iteration evaluationScenarioIteration = evaluationScenarioDatabase.getIteration(iteration);
                System.out.println("\n====================");
                System.out.println("Data for iteration: " + evaluationScenarioIteration.ID);

                System.out.println("\nInitial positions of Agents");
                for (Map.Entry<String, EvaluationScenario.Position> entry : evaluationScenarioIteration.initialPositionsOfAgents.entrySet()) {
                    String agentId = entry.getKey();
                    EvaluationScenario.Position agentInitPos = entry.getValue();
                    System.out.println(
                            "Agent: " + agentId + " x: " + agentInitPos.x + ", y: " + agentInitPos.y + ", z: " + agentInitPos.z
                    );
                }

                System.out.println("\nRanging samples between Agents");
                System.out.println(evaluationScenarioIteration.rangingSamplesBetweenAgents);

                System.out.println("\nRanging samples to Anchors");
                System.out.println(evaluationScenarioIteration.rangingSamplesToAnchors);
            }

        } catch (Exception e) {
            StackTraceElement[] stackTraceElements = e.getStackTrace();
            for (StackTraceElement ste : stackTraceElements) {
                System.out.println(ste.toString());
            }
            System.out.println(e.getLocalizedMessage() + "\nExiting from 1");
        }
    }

    public static void main(String[] args) {
        getNodeSetupFromJsonDB();
    }
}

File: public/content/scripts/EvalScenarioParser.py
import json
from dataclasses import dataclass, field
from typing import List, Dict, Optional, Any
from pathlib import Path


@dataclass
class Position:
    x: float
    y: float
    z: float


@dataclass
class Iteration:
    ID: int
    initial_positions_of_agents: Dict[str, Position] = field(default_factory=dict)
    ranging_samples_between_agents: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    ranging_samples_to_anchors: Dict[str, Dict[str, Any]] = field(default_factory=dict)


@dataclass
class EvaluationScenario:
    Agents: List[int]
    Anchors: List[Any]
    true_positions_of_agents: Dict[str, Position] = field(default_factory=dict)
    true_positions_of_anchors: Dict[str, Position] = field(default_factory=dict)
    Iterations: List[Iteration] = field(default_factory=list)

    def get_iteration(self, iteration_id: int) -> Optional[Iteration]:
        """Retrieve an iteration by ID."""
        for iteration in self.Iterations:
            if iteration.ID == iteration_id:
                return iteration
        return None


def get_node_setup_from_json_db(db_file_path: str) -> None:
    """Parse and display node setup information from JSON database."""
    try:
        print("Parsing Node JSON DB")

        # Read the JSON file
        json_string = Path(db_file_path).read_text()
        data = json.loads(json_string)

        # Deserialize JSON into EvaluationScenario
        evaluation_scenario = EvaluationScenario(
            Agents=data["Agents"],
            Anchors=data["Anchors"],
            true_positions_of_agents={
                k: Position(**v) for k, v in data["True Positions of Agents"].items()
            },
            true_positions_of_anchors={
                k: Position(**v) for k, v in data["True Positions of Anchors"].items()
            },
            Iterations=[
                Iteration(
                    ID=iteration["ID"],
                    initial_positions_of_agents={
                        k: Position(**v) for k, v in iteration["Initial positions of Agents"].items()
                    },
                    ranging_samples_between_agents=iteration.get("Ranging samples between Agents", {}),
                    ranging_samples_to_anchors=iteration.get("Ranging samples to Anchors", {})
                )
                for iteration in data["Iterations"]
            ]
        )

        print("\nNodes in scenario:")
        print(evaluation_scenario.Agents)
        print(evaluation_scenario.Anchors)

        print("\nTrue Positions of Agents")
        for node_id, node_pos in evaluation_scenario.true_positions_of_agents.items():
            print(f"Agent: {node_id} x: {node_pos.x}, y: {node_pos.y}, z: {node_pos.z}")

        print("\nTrue Positions of Anchors")
        for node_id, node_pos in evaluation_scenario.true_positions_of_anchors.items():
            print(f"Anchor: {node_id} x: {node_pos.x}, y: {node_pos.y}, z: {node_pos.z}")

        # Display iteration data
        for iteration_id in range(1, 3):
            iteration = evaluation_scenario.get_iteration(iteration_id)
            if iteration:
                print("\n====================")
                print(f"Data for iteration: {iteration.ID}")

                print("\nInitial positions of Agents")
                for agent_id, agent_init_pos in iteration.initial_positions_of_agents.items():
                    print(f"Agent: {agent_id} x: {agent_init_pos.x}, y: {agent_init_pos.y}, z: {agent_init_pos.z}")

                print("\nRanging samples between Agents")
                print(iteration.ranging_samples_between_agents)

                print("\nRanging samples to Anchors")
                print(iteration.ranging_samples_to_anchors)
            else:
                print(f"Iteration with ID {iteration_id} not found")

    except Exception as e:
        print("An error occurred:", e)
        print("Exiting due to an exception")


if __name__ == "__main__":
    db_file_path = "/Users/geogouz/Desktop/eval scenarios/Local/TWR-CLOUD Bern Eval Scenarios/6905.json"
    get_node_setup_from_json_db(db_file_path)
